json/src/machine.nr

```diff
@@ -0,0 +1,500 @@
+    }
+
+    // Prepare output
+    let mut next_stack: [[Field; 2]; MAX_STACK_HEIGHT] = [[Field::from(0); 2]; MAX_STACK_HEIGHT];
+    let mut next_tree_hash: [[Field; 2]; MAX_STACK_HEIGHT] =
+        [[Field::from(0); 2]; MAX_STACK_HEIGHT];
+
+    for i in 0..MAX_STACK_HEIGHT {
+        next_stack[i][0] = stack[i][0] + (Field::from(indicator[i] as u64) * stack_change_value[0]);
+        let second_index_clear =
+            stack[i][1] * Field::from((read_end_brace | read_end_bracket) as u64);
+        next_stack[i][1] = stack[i][1]
+            + (Field::from(indicator[i] as u64) * (stack_change_value[1] - second_index_clear));
+
+        if to_update_hash[0][i][0] {
+            next_tree_hash[i][0] = tree_hash_change_value[0];
+        } else {
+            next_tree_hash[i][0] = tree_hash[i][0];
+        }
+
+        if to_update_hash[0][i][1] {
+            next_tree_hash[i][1] = tree_hash_change_value[1];
+        } else {
+            next_tree_hash[i][1] = tree_hash[i][1];
+        }
+    }
+
+    RewriteStackOutput { next_monomial, next_stack, next_tree_hash }
+}
+
+//Update current status based on input byte and output the next
+pub fn state_update_hasher(input: StateUpdateHasherInput) -> StateUpdateHasherOutput {
+    let byte = input.byte;
+    let stack = input.stack;
+    let parsing_string = input.parsing_string;
+    let parsing_primitive = input.parsing_primitive;
+    let polynomial_input = input.polynomial_input;
+    let monomial = input.monomial;
+    let tree_hash = input.tree_hash;
+    let escaped = input.escaped;
+
+    // Break down what was read
+    let read_start_brace = is_equal(byte, START_BRACE);
+    let read_end_brace = is_equal(byte, END_BRACE);
+    let read_start_bracket = is_equal(byte, START_BRACKET);
+    let read_end_bracket = is_equal(byte, END_BRACKET);
+    let read_colon = is_equal(byte, COLON);
+    let read_comma = is_equal(byte, COMMA);
+
+    // Read in some delimiter
+    let read_delimiter = read_start_brace
+        | read_end_brace
+        | read_start_bracket
+        | read_end_bracket
+        | read_colon
+        | read_comma;
+
+    // Read in some primitive
+    let primitive_chars: [u8; 23] = [
+        ZERO, ONE, TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE, N, U, L, F, A, S, E, T, R,
+        PERIOD, E_CHAR, PLUS, MINUS,
+    ];
+
+    let read_primitive = contains(byte, primitive_chars);
+
+    // Read in a quote
+    let read_quote = is_equal(byte, QUOTE);
+
+    // Read in an escape
+    let read_escape = is_equal(byte, ESCAPE);
+
+    // Define command instructions
+    let start_brace_instruction = [1, 0, 0];
+    let end_brace_instruction = [2, 0, 0];
+    let start_bracket_instruction = [3, 0, 0];
+    let end_bracket_instruction = [4, 0, 0];
+    let colon_instruction = [5, 0, 0];
+    let comma_instruction = [6, 0, 0];
+    let number_instruction = [0, 0, 1];
+    let quote_instruction = [0, 1, 0];
```

Comment:
These commands are based on this file.
https://github.com/pluto/web-prover-circuits/blob/main/circuits/json/language.circom#L58-L68
```suggestion
    let start_brace_instruction = [1, 0, 0];
    let end_brace_instruction = [-1, 0, -1];
    let start_bracket_instruction = [2, 0, 0];
    let end_bracket_instruction = [-2, 0, -1];
    let quote_instruction = [0, 1, 0];
    let colon_instruction = [3, 0, 0];
    let comma_instruction = [4, 0, -1];
    let number_instruction = [256, 0, 1];
```